### PR TITLE
Salesperson API

### LINKED ISSUE #26806: [BC Idea]: API Page for Sales Persons
### BC Idea Link

https://experience.dynamics.com/ideas/idea/?ideaid=b09d0e1c-3927-ef11-8ee7-6045bdb88c35

### Description

We need an API page for SalesPersons. This page might contain two or more fields - need input if some more fields could be added to this page. Also, I guess this page can be editable, but I am open for discussion. We need it in read-only mode but perhaps for more generic purposes it makes sense to have the page editable.
image

### I will provide the implementation for this BC Idea

- [X] I will provide the implementation for this BC Idea
Internal work item: [AB#541735](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/541735)

### PR DESCRIPTION
**Pull request by miljance**
#### Summary
A new API page for Salespeople/Purchasers
I created the tests in a similar manner that I found suitable for Locations, but I did not manage to run the test. Thus, looking forward to the feedback if tests do pass in Microsoft Pipelines or how to test those calls in a container. Currently I do not have a clue how to do it.

#### Work Item(s) 
Fixes #26806

Fixes [AB#541735](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/541735)

### CODE REVIEW

File: Apps/W1/APIV2/app/src/pages/APIV2SalespersonPurchaser.Page.al [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                }
+                field(code; Rec."Code")
+                {
+                    Caption = 'Code';
+                }
+                field(displayName; Rec.Name)
+                {
+                    Caption = 'Name';
+                }
+                field(eMail; Rec."E-Mail")
```
Comment (onbuyuka) [inline Apps/W1/APIV2/app/src/pages/APIV2SalespersonPurchaser.Page.al:37]:
```suggestion
                field(email; Rec."E-Mail")
```

**Comment by onbuyuka** [comment]
For tests, you need to use Isolation Disabled test runner codeunit

File: Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al [hunk 2]
```diff
@@ -0,0 +1,122 @@
+codeunit 139870 "APIV2 - Salesperson/Purch E2E"
```
Comment (onbuyuka) [inline Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al:1]: It'd be great to add an Update test too

**Review by pri-kise — APPROVED** [review]

**Review by onbuyuka — DISMISSED** [review]

**Review by JesperSchulz — DISMISSED** [review]

File: Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
```
Comment (JesperSchulz) [inline Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al:1]:
```suggestion
codeunit 139882 "APIV2 - Salesperson/Purch E2E"
```

**Review by JesperSchulz — DISMISSED** [review]

**Review by onbuyuka — DISMISSED** [review]

File: Apps/W1/APIV2/app/src/pages/APIV2SalespersonPurchaser.Page.al [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                }
+                field(coupledToDataverse; Rec."Coupled to Dataverse")
+                {
+                    Caption = 'Coupled to Dataverse';
+                }
+                field(lastModifiedDateTime; Rec.SystemModifiedAt)
+                {
+                    Caption = 'Last Modified Date';
+                    Editable = false;
+                }
```
Comment (JesperSchulz) [inline Apps/W1/APIV2/app/src/pages/APIV2SalespersonPurchaser.Page.al:73]:
```suggestion
                field(blocked; Rec.Blocked)
                {
                    Caption = 'Blocked';
                }
                field(lastModifiedDateTime; Rec.SystemModifiedAt)
                {
                    Caption = 'Last Modified Date';
                    Editable = false;
                }
```

File: Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'email2', SalespersonPurchaser."E-Mail 2");
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'phoneNo', SalespersonPurchaser."Phone No.");
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'jobTitle', SalespersonPurchaser."Job Title");
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'commisionPercent', SalespersonPurchaser."Commission %");
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'privacyBlocked', SalespersonPurchaser."Privacy Blocked");
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'blocked', SalespersonPurchaser.Blocked);
+        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'coupledToDataverse', SalespersonPurchaser."Coupled to Dataverse");
+        exit(SalespersonPurchaserJSON);
```
Comment (JesperSchulz) [inline Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al:156]:
```suggestion
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'blocked', SalespersonPurchaser.Blocked);
        exit(SalespersonPurchaserJSON);
```

File: Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'email', SalespersonPurchaser."E-Mail");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'email2', SalespersonPurchaser."E-Mail 2");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'phoneNo', SalespersonPurchaser."Phone No.");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'jobTitle', SalespersonPurchaser."Job Title");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'commisionPercent', Format(SalespersonPurchaser."Commission %", 0, 9));
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'privacyBlocked', Format(SalespersonPurchaser."Privacy Blocked", 0, 9));
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'blocked', Format(SalespersonPurchaser.Blocked, 0, 9));
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'coupledToDataverse', Format(SalespersonPurchaser."Coupled to Dataverse", 0, 9));
+    end;
+}
```
Comment (JesperSchulz) [inline Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al:174]:
```suggestion
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'jobTitle', SalespersonPurchaser."Job Title");
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'commisionPercent', Format(SalespersonPurchaser."Commission %", 0, 9));
    end;
}
```

**Comment by JesperSchulz** [comment]
This one is merged internally. Not sure why the tests are failing here. I'm inclined just to close this PR. The changes will get pushed to this repo when we push out the next code update.

File: Apps/W1/APIV2/app/src/pages/APIV2SalespersonPurchaser.Page.al [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        }
+    }
+}
```
Comment (JesperSchulz) [inline Apps/W1/APIV2/app/src/pages/APIV2SalespersonPurchaser.Page.al:73]: Making sure the code is identical to the code we checked in internally.

```suggestion
namespace Microsoft.API.V2;

using Microsoft.CRM.Team;

page 30075 "APIV2 - Salesperson/Purchaser"
{
    DelayedInsert = true;
    APIVersion = 'v2.0';
    EntityCaption = 'Salesperson/Purchaser';
    EntitySetCaption = 'Salespeople/Purchasers';
    PageType = API;
    ODataKeyFields = SystemId;
    EntityName = 'salespersonPurchaser';
    EntitySetName = 'salespeoplePurchasers';
    SourceTable = "Salesperson/Purchaser";
    Extensible = false;

    layout
    {
        area(Content)
        {
            repeater(Group)
            {
                field(id; Rec.SystemId)
                {
                    Caption = 'Id';
                    Editable = false;
                }
                field(code; Rec."Code")
                {
                    Caption = 'Code';
                }
                field(displayName; Rec.Name)
                {
                    Caption = 'Name';
                }
                field(email; Rec."E-Mail")
                {
                    Caption = 'Email';
                }
                field(email2; Rec."E-Mail 2")
                {
                    Caption = 'Email 2';
                }
                field(phoneNo; Rec."Phone No.")
                {
                    Caption = 'Phone No.';
                }
                field(jobTitle; Rec."Job Title")
                {
                    Caption = 'Job Title';
                }
                field(commisionPercent; Rec."Commission %")
                {
                    Caption = 'Commission %';
                }
                field(privacyBlocked; Rec."Privacy Blocked")
                {
                    Caption = 'Privacy Blocked';
                }
                field(blocked; Rec.Blocked)
                {
                    Caption = 'Blocked';
                }
                field(lastModifiedDateTime; Rec.SystemModifiedAt)
                {
                    Caption = 'Last Modified Date';
                    Editable = false;
                }
            }
        }
    }
}
```

File: Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        LibraryGraphMgt.VerifyIDInJSON(JSON);
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'code', SalespersonPurchaser.Code);
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'displayName', SalespersonPurchaser.Name);
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'email', SalespersonPurchaser."E-Mail");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'email2', SalespersonPurchaser."E-Mail 2");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'phoneNo', SalespersonPurchaser."Phone No.");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'jobTitle', SalespersonPurchaser."Job Title");
+        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'commisionPercent', Format(SalespersonPurchaser."Commission %", 0, 9));
+    end;
+}
```
Comment (JesperSchulz) [inline Apps/W1/APIV2/test/src/APIV2SalespersonPurchE2E.Codeunit.al:170]: Making sure the code is identical to the code we checked in internally.
```suggestion
codeunit 139882 "APIV2 - Salesperson/Purch E2E"
{
    Subtype = Test;

    var
        Assert: Codeunit Assert;
        LibraryGraphMgt: Codeunit "Library - Graph Mgt";
        LibraryUtility: Codeunit "Library - Utility";
        LibraryRandom: Codeunit "Library - Random";
        IsInitialized: Boolean;
        EmptyJSONErr: Label 'JSON should not be empty.', Locked = true;
        ServiceNameTxt: Label 'salespeoplePurchasers', Locked = true;

    local procedure Initialize()
    begin
        if IsInitialized then
            exit;

        IsInitialized := true;
        Commit();
    end;

    [Test]
    procedure GetSalespersonPurchaser()
    var
        SalespersonPurchaser: Record "Salesperson/Purchaser";
        Response: Text;
        TargetURL: Text;
    begin
        // [SCENARIO] User can get a salesperson/purchaser with a GET request to the service.
        Initialize();

        // [GIVEN] A salesperson/purchaser exists in the system.
        CreateSalespersonPurchaser(SalespersonPurchaser);

        // [WHEN] The user makes a GET request for a given salesperson/purchaser.
        TargetURL := LibraryGraphMgt.CreateTargetURL(SalespersonPurchaser.SystemId, Page::"APIV2 - Salesperson/Purchaser", ServiceNameTxt);
        LibraryGraphMgt.GetFromWebService(Response, TargetURL);

        // [THEN] The response text contains the information about salesperson/purchaser.
        VerifyProperties(Response, SalespersonPurchaser);
    end;

    [Test]
    procedure CreateSalespersonPurchaser()
    var
        SalespersonPurchaser: Record "Salesperson/Purchaser";
        TempSalespersonPurchaser: Record "Salesperson/Purchaser" temporary;
        SalespersonPurchaserJSON: Text;
        TargetURL: Text;
        Response: Text;
    begin
        // [SCENARIO] User can create a new salesperson/purchaser through a POST method.
        Initialize();

        // [GIVEN] The user has constructed a salesperson/purchaser JSON object to send to the service.
        CreateSalespersonPurchaser(TempSalespersonPurchaser);
        SalespersonPurchaserJSON := GetSalespersonPurchaserJSON(TempSalespersonPurchaser);

        // [WHEN] The user posts the JSON to the service.
        TargetURL := LibraryGraphMgt.CreateTargetURL('', Page::"APIV2 - Salesperson/Purchaser", ServiceNameTxt);
        LibraryGraphMgt.PostToWebService(TargetURL, SalespersonPurchaserJSON, Response);

        // [THEN] The salesperson/purchaser has been created in the database with all the details.
        SalespersonPurchaser.Get(TempSalespersonPurchaser.Code);
        VerifyProperties(Response, SalespersonPurchaser);
    end;

    [Test]
    procedure ModifySalespersonPurchaserJobTitle()
    var
        SalespersonPurchaser: Record "Salesperson/Purchaser";
        TempSalespersonPurchaser: Record "Salesperson/Purchaser" temporary;
        RequestBody: Text;
        Response: Text;
        TargetURL: Text;
    begin
        // [SCENARIO] User can modify a salesperson/purchaser through a PATCH request.
        Initialize();

        // [GIVEN] A salesperson/purchaser exists.
        CreateSalespersonPurchaser(SalespersonPurchaser);
        TempSalespersonPurchaser.TransferFields(SalespersonPurchaser);
        TempSalespersonPurchaser."Job Title" := LibraryUtility.GenerateGUID();
        RequestBody := GetSalespersonPurchaserJSON(TempSalespersonPurchaser);

        // [WHEN] The user makes a patch request to the service and specifies Job Title field.
        TargetURL := LibraryGraphMgt.CreateTargetURL(SalespersonPurchaser.SystemId, Page::"APIV2 - Salesperson/Purchaser", ServiceNameTxt);
        LibraryGraphMgt.PatchToWebService(TargetURL, RequestBody, Response);

        // [THEN] The response contains the new values.
        VerifyProperties(Response, TempSalespersonPurchaser);

        // [THEN] The salesperson/purchaser in the database contains the updated values.
        SalespersonPurchaser.Get(SalespersonPurchaser.Code);
        Assert.AreEqual(SalespersonPurchaser."Job Title", TempSalespersonPurchaser."Job Title", 'Job Title should be equal.');
    end;

    [Test]
    procedure DeleteSalespersonPurchaser()
    var
        SalespersonPurchaser: Record "Salesperson/Purchaser";
        SalespersonPurchaserCode: Code[10];
        TargetURL: Text;
        Response: Text;
    begin
        // [SCENARIO] User can delete a salesperson/purchaser by making a DELETE request.
        Initialize();

        // [GIVEN] A salesperson/purchaser exists.
        CreateSalespersonPurchaser(SalespersonPurchaser);
        SalespersonPurchaserCode := SalespersonPurchaser.Code;

        // [WHEN] The user makes a DELETE request to the endpoint for the salesperson/purchaser.
        TargetURL := LibraryGraphMgt.CreateTargetURL(SalespersonPurchaser.SystemId, Page::"APIV2 - Salesperson/Purchaser", ServiceNameTxt);
        LibraryGraphMgt.DeleteFromWebService(TargetURL, '', Response);

        // [THEN] The response is empty.
        Assert.AreEqual('', Response, 'DELETE response should be empty.');

        // [THEN] The salesperson/purchaser is no longer in the database.
        SalespersonPurchaser.SetRange(Code, SalespersonPurchaserCode);
        Assert.IsTrue(SalespersonPurchaser.IsEmpty(), 'Salesperson/Purchaser should be deleted.');
    end;

    local procedure CreateSalespersonPurchaser(var SalespersonPurchaser: Record "Salesperson/Purchaser"): Code[20]
    begin
        SalespersonPurchaser.Init();
        SalespersonPurchaser.Validate(
          Code, LibraryUtility.GenerateRandomCode(SalespersonPurchaser.FieldNo(Code), DATABASE::"Salesperson/Purchaser"));
        SalespersonPurchaser.Name := LibraryUtility.GenerateRandomText(50);
        SalespersonPurchaser."E-Mail" := LibraryUtility.GenerateRandomEmail();
        SalespersonPurchaser."E-Mail 2" := LibraryUtility.GenerateRandomEmail();
        SalespersonPurchaser."Phone No." := LibraryUtility.GenerateRandomPhoneNo();
        SalespersonPurchaser."Job Title" := LibraryUtility.GenerateRandomText(30);
        SalespersonPurchaser."Commission %" := LibraryRandom.RandDec(100, 2);
        SalespersonPurchaser."Privacy Blocked" := true;
        SalespersonPurchaser.Blocked := true;
        SalespersonPurchaser."Coupled to Dataverse" := true;
        SalespersonPurchaser.Insert(true);
        Commit();
        exit(SalespersonPurchaser.Code);
    end;

    local procedure GetSalespersonPurchaserJSON(var SalespersonPurchaser: Record "Salesperson/Purchaser") SalespersonPurchaserJSON: Text
    begin
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'code', SalespersonPurchaser.Code);
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'displayName', SalespersonPurchaser.Name);
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'email', SalespersonPurchaser."E-Mail");
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'email2', SalespersonPurchaser."E-Mail 2");
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'phoneNo', SalespersonPurchaser."Phone No.");
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'jobTitle', SalespersonPurchaser."Job Title");
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'commisionPercent', SalespersonPurchaser."Commission %");
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'privacyBlocked', SalespersonPurchaser."Privacy Blocked");
        SalespersonPurchaserJSON := LibraryGraphMgt.AddPropertytoJSON(SalespersonPurchaserJSON, 'blocked', SalespersonPurchaser.Blocked);
        exit(SalespersonPurchaserJSON);
    end;

    local procedure VerifyProperties(JSON: Text; SalespersonPurchaser: Record "Salesperson/Purchaser")
    begin
        Assert.AreNotEqual('', JSON, EmptyJSONErr);
        LibraryGraphMgt.VerifyIDInJSON(JSON);
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'code', SalespersonPurchaser.Code);
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'displayName', SalespersonPurchaser.Name);
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'email', SalespersonPurchaser."E-Mail");
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'email2', SalespersonPurchaser."E-Mail 2");
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'phoneNo', SalespersonPurchaser."Phone No.");
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'jobTitle', SalespersonPurchaser."Job Title");
        LibraryGraphMgt.VerifyPropertyInJSON(JSON, 'commisionPercent', Format(SalespersonPurchaser."Commission %", 0, 9));
    end;
}
```

**Review by JesperSchulz — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Tests probably failing due to setup issue. We will have to investigate.

**Comment by JesperSchulz** [comment]
@aholstrup1 will force merge and update the repo to get this in.

**Pull request merged**